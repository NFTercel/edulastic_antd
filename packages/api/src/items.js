import API from './utils/API';

const api = new API();

const receiveItems = () =>
  api.callApi({ url: '/items' }).then(result => result.data);

const receiveItemById = id =>
  api
    .callApi({
      url: `/items/${id}`
    })
    .then(result => result.data);

const createItem = ({ payload }) =>
  api
    .callApi({
      method: 'post',
      url: '/items',
      data: {
        ...payload
      }
    })
    .then(result => result.data);

const updateItemById = ({ payload }) =>
  api
    .callApi({
      method: 'put',
      url: `/items/${payload.id}`,
      data: {
        ...payload
      }
    })
    .then(result => result.data);

const saveUserReponse = data =>
  api
    .callApi({
      method: 'post',
      url: '/UserTestItemActivities/saveResponse',
      data
    })
    .then(result => result.data);

const getUserResponse = itemId =>
  api
    .callApi({
      method: 'get',
      url: 'UserTestItemActivities',
      params: { item: itemId }
    })
    .then(result => result.data);

export default {
  receiveItems,
  receiveItemById,
  createItem,
  updateItemById,
  saveUserReponse,
  getUserResponse
};
