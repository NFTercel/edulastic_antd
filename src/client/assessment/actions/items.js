import {
  RECEIVE_ITEMS_REQUEST,
  RECEIVE_ITEM_REQUEST,
  CREATE_ITEM_REQUEST,
  UPDATE_ITEM_REQUEST,
  GOTO_ITEM,
  SAVE_USER_RESPONSE,
  LOAD_USER_RESPONSE
} from "../constants/actions";

export const receiveItemsAction = ({ page, limit, search }) => ({
  type: RECEIVE_ITEMS_REQUEST,
  payload: { page, limit, search }
});

export const receiveItemByIdAction = id => ({
  type: RECEIVE_ITEM_REQUEST,
  payload: { id }
});

export const createItemAction = payload => ({
  type: CREATE_ITEM_REQUEST,
  payload: { payload }
});

export const updateItemByIdAction = payload => ({
  type: UPDATE_ITEM_REQUEST,
  payload: { payload }
});

export const gotoItem = item => ({
  type: GOTO_ITEM,
  payload: {
    item
  }
});

export const saveUserResponse = item => ({
  type: SAVE_USER_RESPONSE,
  payload: {
    itemId: item
  }
});

export const loadUserResponse = item => ({
  type: LOAD_USER_RESPONSE,
  payload: {
    itemId: item
  }
});
