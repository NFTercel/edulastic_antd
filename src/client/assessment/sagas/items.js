import { takeLatest, call, put, all, select } from "redux-saga/effects";
import { itemsApi, testItemActivityApi } from "@edulastic/api";
import { getCurrentGroup } from "../../student/Login/ducks";
import {
  RECEIVE_ITEM_REQUEST,
  RECEIVE_ITEM_SUCCESS,
  RECEIVE_ITEM_ERROR,
  RECEIVE_ITEMS_SUCCESS,
  RECEIVE_ITEMS_ERROR,
  SAVE_USER_RESPONSE,
  LOAD_USER_RESPONSE,
  LOAD_ANSWERS
} from "../constants/actions";

function* receiveItemsSaga() {
  try {
    const items = yield call(itemsApi.receiveItems);

    yield put({
      type: RECEIVE_ITEMS_SUCCESS,
      payload: { items }
    });
  } catch (err) {
    console.error(err);
    yield put({
      type: RECEIVE_ITEMS_ERROR,
      payload: { error: "Receive items is failing" }
    });
  }
}

function* receiveItemSaga({ payload }) {
  try {
    const item = yield call(itemsApi.receiveItemById, payload.id);

    yield put({
      type: RECEIVE_ITEM_SUCCESS,
      payload: { item }
    });
  } catch (err) {
    console.error(err);
    yield put({
      type: RECEIVE_ITEM_ERROR,
      payload: { error: "Receive item by id is failing" }
    });
  }
}

// fetch all questionIds from item
const getQuestionIds = item => {
  let questions = [];
  item.rows.forEach(row => {
    questions = [...questions, ...row.widgets.map(widget => widget.reference)].filter(q => !!q);
  });

  return questions;
};

function* saveUserResponse({ payload }) {
  try {
    const itemIndex = payload.itemId;
    const items = yield select(state => state.test && state.test.items);
    const answers = yield select(state => state.answers);
    const userTestActivityId = yield select(state => state.test && state.test.testActivityId);

    const currentItem = items.length && items[itemIndex];
    const questions = getQuestionIds(currentItem);
    const itemAnswers = {};
    questions.forEach(question => {
      itemAnswers[question] = answers[question];
    });

    const testItemId = currentItem._id;
    const assignmentId = yield select(state => state.studentAssignment && state.studentAssignment.current);
    const userWork = yield select(({ userWork }) => userWork[testItemId]);
    const groupId = yield select(getCurrentGroup);

    const activity = {
      answers: itemAnswers,
      testItemId,
      assignmentId,
      testActivityId: userTestActivityId,
      groupId
    };
    if (userWork) activity.userWork = userWork;

    yield call(testItemActivityApi.create, activity);
  } catch (err) {
    console.log(err);
  }
}

function* loadUserResponse({ payload }) {
  try {
    const itemIndex = payload.itemId;
    const items = yield select(state => state.test && state.test.items);
    const item = items[itemIndex];
    const { answers } = yield call(itemsApi.getUserResponse, item._id);
    yield put({
      type: LOAD_ANSWERS,
      payload: {
        ...answers
      }
    });
  } catch (e) {
    console.log(e);
  }
}
export default function* watcherSaga() {
  yield all([
    yield takeLatest(RECEIVE_ITEM_REQUEST, receiveItemSaga),
    yield takeLatest(SAVE_USER_RESPONSE, saveUserResponse),
    yield takeLatest(LOAD_USER_RESPONSE, loadUserResponse)
  ]);
}
