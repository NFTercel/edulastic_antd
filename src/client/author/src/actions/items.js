import {
  RECEIVE_ITEMS_REQUEST,
  RECEIVE_ITEM_REQUEST,
  CREATE_ITEM_REQUEST,
  UPDATE_ITEM_REQUEST
} from "../constants/actions";

export const receiveItemsAction = ({ page, limit, search }) => ({
  type: RECEIVE_ITEMS_REQUEST,
  payload: { page, limit, search }
});

export const receiveItemByIdAction = id => ({
  type: RECEIVE_ITEM_REQUEST,
  payload: { id }
});

export const createItemAction = payload => ({
  type: CREATE_ITEM_REQUEST,
  payload: { payload }
});

export const updateItemByIdAction = payload => ({
  type: UPDATE_ITEM_REQUEST,
  payload: { payload }
});
