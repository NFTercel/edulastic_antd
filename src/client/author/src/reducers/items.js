import { combineReducers } from "redux";

import {
  RECEIVE_ITEMS_REQUEST,
  RECEIVE_ITEMS_SUCCESS,
  RECEIVE_ITEMS_ERROR,
  RECEIVE_ITEM_REQUEST,
  RECEIVE_ITEM_SUCCESS,
  RECEIVE_ITEM_ERROR
} from "../constants/actions";

const initialItemsState = {
  items: [],
  error: null,
  loading: false,
  page: 1,
  limit: 5,
  count: 0
};

export const itemsReducer = (state = initialItemsState, { type, payload }) => {
  switch (type) {
    case RECEIVE_ITEMS_REQUEST:
      return { ...state, loading: true };
    case RECEIVE_ITEMS_SUCCESS:
      return {
        ...state,
        loading: false,
        items: payload.items,
        page: payload.page,
        limit: payload.limit,
        count: payload.count
      };
    case RECEIVE_ITEMS_ERROR:
      return { ...state, loading: false, error: payload.error };
    default:
      return state;
  }
};

const initialItemState = {
  item: null,
  error: null,
  loading: false
};

export const itemReducer = (state = initialItemState, { type, payload }) => {
  switch (type) {
    case RECEIVE_ITEM_REQUEST:
      return { ...state, loading: true };
    case RECEIVE_ITEM_SUCCESS:
      return { ...state, loading: false, item: payload.item };
    case RECEIVE_ITEM_ERROR:
      return { ...state, loading: false, error: payload.error };
    default:
      return state;
  }
};

export default combineReducers({
  list: itemsReducer,
  item: itemReducer
});
