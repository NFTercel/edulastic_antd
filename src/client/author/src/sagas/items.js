import { takeEvery, takeLatest, call, put, all } from "redux-saga/effects";
import { itemsApi } from "@edulastic/api";
import { message } from "antd";

import {
  RECEIVE_ITEM_REQUEST,
  RECEIVE_ITEM_SUCCESS,
  RECEIVE_ITEM_ERROR,
  RECEIVE_ITEMS_REQUEST,
  RECEIVE_ITEMS_SUCCESS,
  RECEIVE_ITEMS_ERROR,
  CREATE_ITEM_REQUEST,
  UPDATE_ITEM_REQUEST
} from "../constants/actions";

export function* receiveItemsSaga({ payload }) {
  try {
    const { items, page, count } = yield call(itemsApi.receiveItems, payload);

    yield put({
      type: RECEIVE_ITEMS_SUCCESS,
      payload: { items, page, limit: payload.limit, count }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Receive items is failing";
    yield call(message.error, errorMessage);
    yield put({
      type: RECEIVE_ITEMS_ERROR,
      payload: { error: errorMessage }
    });
  }
}

export function* receiveItemSaga({ payload }) {
  try {
    const item = yield call(itemsApi.receiveItemById, payload.id);

    yield put({
      type: RECEIVE_ITEM_SUCCESS,
      payload: { item }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Receive item by id is failing";
    yield call(message.error, errorMessage);
    yield put({
      type: RECEIVE_ITEM_ERROR,
      payload: { error: errorMessage }
    });
  }
}

export function* createItemSaga({ payload }) {
  try {
    const item = yield call(itemsApi.createItem, payload);
    yield put({
      type: RECEIVE_ITEM_SUCCESS,
      payload: { item: item.data }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Create item is failed";
    yield call(message.error, errorMessage);
    yield put({
      type: RECEIVE_ITEM_ERROR,
      payload: { error: errorMessage }
    });
  }
}

export function* updateItemSaga({ payload }) {
  try {
    const item = yield call(itemsApi.updateItemById, payload);
    yield put({
      type: RECEIVE_ITEM_SUCCESS,
      payload: { item: item.data }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Update item is failed";
    yield call(message.error, errorMessage);
    yield put({
      type: RECEIVE_ITEM_ERROR,
      payload: { error: errorMessage }
    });
  }
}

export default function* watcherSaga() {
  yield all([
    yield takeEvery(RECEIVE_ITEM_REQUEST, receiveItemSaga),
    yield takeLatest(RECEIVE_ITEMS_REQUEST, receiveItemsSaga),
    yield takeLatest(CREATE_ITEM_REQUEST, createItemSaga),
    yield takeLatest(UPDATE_ITEM_REQUEST, updateItemSaga)
  ]);
}
