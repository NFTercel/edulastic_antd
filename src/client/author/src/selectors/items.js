import { createSelector } from "reselect";

export const stateSelector = state => state.items;
export const itemsSelector = createSelector(
  stateSelector,
  state => state.list
);
export const itemSelector = createSelector(
  stateSelector,
  state => state.item
);

export const getItemSelector = createSelector(
  itemSelector,
  state => state.item
);
export const getItemLoadingSelector = createSelector(
  itemSelector,
  state => state.loading
);

export const getItemsListSelector = createSelector(
  itemsSelector,
  state => state.items
);
export const getItemsPageSelector = createSelector(
  itemsSelector,
  state => state.page
);
export const getItemsLimitSelector = createSelector(
  itemsSelector,
  state => state.limit
);
export const getItemsCountSelector = createSelector(
  itemsSelector,
  state => state.count
);
export const getItemsLoadingSelector = createSelector(
  itemsSelector,
  state => state.loading
);
