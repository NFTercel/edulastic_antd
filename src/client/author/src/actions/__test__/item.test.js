import { receiveItemsAction, receiveItemByIdAction, createItemAction, updateItemByIdAction } from "../items";

describe("Items", () => {
  it("receive items should return an action", () => {
    expect(
      receiveItemsAction({
        page: 1,
        limit: 5,
        search: "items"
      })
    ).toMatchSnapshot();
  });

  it("receive item by id should return an action", () => {
    expect(
      receiveItemByIdAction({
        id: 5
      })
    ).toMatchSnapshot();
  });

  it("receive item by id should return an action", () => {
    expect(
      createItemAction({
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: [
              {
                widgetType: "question",
                type: "multipleChoice",
                title: "Multiple choice",
                reference: "5c0f0c9ed8c7f4453f5bb51f",
                tabIndex: 0
              }
            ]
          }
        ],
        columns: []
      })
    ).toMatchSnapshot();
  });

  it("receive item by id should return an action", () => {
    expect(
      updateItemByIdAction({
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: [
              {
                widgetType: "question",
                type: "multipleChoice",
                title: "Multiple choice",
                reference: "5c0f0c9ed8c7f4453f5bb51f",
                tabIndex: 0
              }
            ]
          }
        ],
        columns: []
      })
    ).toMatchSnapshot();
  });
});
