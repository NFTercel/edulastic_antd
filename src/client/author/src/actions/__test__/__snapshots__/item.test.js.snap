// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items receive item by id should return an action 1`] = `
Object {
  "payload": Object {
    "id": Object {
      "id": 5,
    },
  },
  "type": "[items] receive item request",
}
`;

exports[`Items receive item by id should return an action 2`] = `
Object {
  "payload": Object {
    "payload": Object {
      "columns": Array [],
      "rows": Array [
        Object {
          "dimension": "100%",
          "tabs": Array [],
          "widgets": Array [
            Object {
              "reference": "5c0f0c9ed8c7f4453f5bb51f",
              "tabIndex": 0,
              "title": "Multiple choice",
              "type": "multipleChoice",
              "widgetType": "question",
            },
          ],
        },
      ],
    },
  },
  "type": "[items] create item request",
}
`;

exports[`Items receive item by id should return an action 3`] = `
Object {
  "payload": Object {
    "payload": Object {
      "columns": Array [],
      "rows": Array [
        Object {
          "dimension": "100%",
          "tabs": Array [],
          "widgets": Array [
            Object {
              "reference": "5c0f0c9ed8c7f4453f5bb51f",
              "tabIndex": 0,
              "title": "Multiple choice",
              "type": "multipleChoice",
              "widgetType": "question",
            },
          ],
        },
      ],
    },
  },
  "type": "[items] update item request",
}
`;

exports[`Items receive items should return an action 1`] = `
Object {
  "payload": Object {
    "limit": 5,
    "page": 1,
    "search": "items",
  },
  "type": "[items] receive items request",
}
`;
